pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package traps
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/gosnmp/gosnmp"
+)
+
+// Formatter is an interface to extract and format raw SNMP Traps
+type Formatter interface {
+	FormatPacket(packet *SnmpPacket) ([]byte, error)
+	GetTags(packet *SnmpPacket) []string
+}
+
+// JSONFormatter is a Formatter implementation that transforms Traps into JSON
+type JSONFormatter struct {
+	oidResolver OIDResolver
+	namespace   string
+}
+
+type trapVariable struct {
+	OID     string      `json:"oid"`
+	VarType string      `json:"type"`
+	Value   interface{} `json:"value"`
+}
+
+const (
+	sysUpTimeInstanceOID = "1.3.6.1.2.1.1.3.0"
+	snmpTrapOID          = "1.3.6.1.6.3.1.1.4.1.0"
+)
+
+// NewJSONFormatter creates a new JSONFormatter instance with an optional OIDResolver variable.
+func NewJSONFormatter(oidResolver OIDResolver) JSONFormatter {
+	namespace := GetNamespace()
+	if oidResolver == nil {
+		return JSONFormatter{NoOpOIDResolver{}, namespace}
```

Comment:
Related to my other question "Should we exit and return an error if we don't have a valid resolver ?"

If we exit earlier, there is maybe no need for a `NoOpOIDResolver`.

---

I'm not sure to understand. If NoOpOIDResolver is used here, it means that if traps are received, the resolver will always return an error if GetTrapMetadata/GetVariableMetadata is called?